src/mlir/lowering.md

```diff
@@ -0,0 +1,517 @@
+
+## Basic Blocks
+
+So far, there is only two constructs that create new basic blocks: conditionals
+and loops. Both will end with a tail block that will be the main block of the
+subsequent operations.
+
+The main issue when lowering basic blocks is figuring out their arguments. For
+this we need free variable and liveness analysis, which will have already been
+done by the 1IR passes. That information is in the `FnAnalysis` structure for
+now and it should be trivial to use it for arguments without having to backtrack.
+
+Once the 1IR becomes more complete (with less side structures), we'll need some
+information in the IR itself (condition/loop nodes).
+
+## Types
+
+All types in 1IR should already be concrete at the time of conversion, so all
+Verona types in the 2IR must validate using standard MLIR checks, which only
+check for equality, not similarity or sub-types.
+
+If there are any cases where this is not true (in != out), we can add casts in
+expectation that the type check in 1IR was successful and therefore the cast is
+valid.
+
+Class types should be lowered to a `!verona.class<fields:types>`. We don't need
+a similar struct representation, since every field read/write should have the
+class type and the field name, with the concrete field type as a result.
+
+## Regions
+
+Every object is allocated into a region. Some objects share regions, others
+live in immutable memory. Whether we'll allow users to annotate objects with
+region information (ex. `new Foo in Bar`) or not is up to debate. Regardless,
+we'll need that kind of information in the IR to add calls to the right runtime
+library functions that control memory.
+
+We want a pass that bundles objects into regions based on their access patterns
+taking user annotations as either hints or requirements. Such pass could be done
+in 1IR or 2IR, but it will depending where it runs and whether we allow user
+annotations will dictate if the 1IR will have region information, but the 2IR
+will definitely need it.
+
+Region informaiton can be represented as references to the owning `iso` (see
+below) on the types on every instance of, to avoid repeated and potentially
+complex search patterns.
+
+### ISO
+
+Iso references are region owning pointers that can freely access the whole region
+and sub-regions (if not given to another task). Any read-write access through
+those references is free.
+
+### MUT
+
+Mutable references must belong to a region and must have access to that region,
+ie. they need to belong to the owning `iso` of the region. For this, mutable
+references must point to the `iso` references. This can be represented as
+MLIR annotation, referring the SSA value (ex. `#region<%22>` where `%22` must be
+an `iso` reference of the same region).
+
+### drop
+
+Dropping (deallocating) regions need to know that there are no more active users
+of that region (ex. reference counting) and also need to find which region it is.
```

Comment:
Regions are deallocated when the `iso` goes out of scope / is overwritten. We don't do any kind of smart trick to find "active users" of it.